src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultCollectionValidationStrategy.cs

```diff
@@ -60,9 +63,16 @@ private DefaultCollectionValidationStrategy()
             return new Enumerator(metadata.ElementMetadata, key, enumerator);
         }
 
-        public static IEnumerator GetEnumeratorForElementType(ModelMetadata metadata, object model)
+        public IEnumerator GetEnumeratorForElementType(ModelMetadata metadata, object model)
         {
-            var getEnumeratorMethod = _getEnumerator.MakeGenericMethod(metadata.ElementType);
+            MethodInfo getEnumeratorMethod;
+            if (!_genericGetEnumeratorCache.TryGetValue(metadata.ElementType, out getEnumeratorMethod))
+            {
+                getEnumeratorMethod = _genericGetEnumeratorCache.AddOrUpdate(
+                        key: metadata.ElementType,
+                        addValueFactory: (type) => _getEnumerator.MakeGenericMethod(type),
+                        updateValueFactory: (type, methodInfo) => methodInfo);
+            }
             return (IEnumerator)getEnumeratorMethod.Invoke(null, new object[] { model });
```

Comment:
Yea, feels like a case where we should compile another expression tree. We're pretty good at that now 😄